### PR TITLE
[#30] Setup network layer

### PR DESCRIPTION
**Pull request by luongvo**
https://github.com/luongvo/flutter-survey/issues/30

## What happened 👀

Before having more backend tasks on some screens, we have to set up the networking layer.
 
## Insight 📝

- Endpoint
  - Production: https://survey-api.nimblehq.co/
  - Staging: https://nimble-survey-web-staging.herokuapp.com/ 
- Setup network components with [Dio](https://pub.dev/packages/dio), [retrofit](https://pub.dev/packages/retrofit) and [freezed](https://pub.dev/packages/freezed)
- Setup DI with [get_it](https://pub.dev/packages/get_it) and [injectable](https://pub.dev/packages/injectable)
- Setup `BaseUseCase` and usecase's `Result`
- Show request log
 
## Proof Of Work 📹

N/A

### CODE REVIEW

File: lib/flavors.dart [hunk 1]
```diff
@@ -0,0 +1,44 @@
+enum Flavor {
```
Comment (manh-t) [inline lib/flavors.dart:1]: @luongvo by using `flutter_config` and setting up `flavor` manually from #2 , we can move these configurations into `.env` file. This `flavors.dart` is redundant now 😁

File: lib/main-production.dart [hunk 2]
```diff
@@ -0,0 +1,7 @@
+import 'flavors.dart';
```
Comment (manh-t) [inline lib/main-production.dart:1]: @luongvo This file is redundant because we don't use `flutter-flavorizr` anymore

File: lib/main-staging.dart [hunk 3]
```diff
@@ -0,0 +1,7 @@
+import 'flavors.dart';
```
Comment (manh-t) [inline lib/main-staging.dart:1]: @luongvo This file is redundant because we don't use `flutter-flavorizr` anymore 🙋

File: lib/flavors.dart (hunk 1 above)
Comment (luongvo) [reply lib/flavors.dart:1]: @manh-t yeah, I'm looking a way to get flavor state in main to set corresponding credentials for the network components. Sounds nice to go, try try 🙆‍♂️

File: lib/api/exception/network_exceptions.dart [hunk 4]
```diff
@@ -0,0 +1,158 @@
...
+            case DioErrorType.receiveTimeout:
+              networkExceptions = NetworkExceptions.sendTimeout();
+              break;
+            case DioErrorType.sendTimeout:
+              networkExceptions = NetworkExceptions.sendTimeout();
+              break;
+            case DioErrorType.response:
+              switch (error.response?.statusCode) {
+                case 400:
+                  networkExceptions = NetworkExceptions.unauthorisedRequest();
```
Comment (sleepylee) [inline lib/api/exception/network_exceptions.dart:68]: badRequest() 🤣

File: lib/api/exception/network_exceptions.dart [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+              networkExceptions = NetworkExceptions.requestCancelled();
+              break;
+            case DioErrorType.connectTimeout:
+              networkExceptions = NetworkExceptions.requestTimeout();
+              break;
+            case DioErrorType.other:
+              networkExceptions = NetworkExceptions.noInternetConnection();
+              break;
+            case DioErrorType.receiveTimeout:
+              networkExceptions = NetworkExceptions.sendTimeout();
```
Comment (sleepylee) [inline lib/api/exception/network_exceptions.dart:60]: receiveTimeout() 🤣  I just noticed as well

File: lib/api/exception/network_exceptions.dart (hunk 4 above)
Comment (luongvo) [reply lib/api/exception/network_exceptions.dart:68]: you suck

Comment (luongvo) [reply lib/api/exception/network_exceptions.dart]: Fixed https://github.com/luongvo/flutter-survey/pull/50/commits/0464ce8e3f93e91170a8d1bf1b484d1e55ff9a24

**Review by sleepylee — APPROVED** [review]

**Review by manh-t — APPROVED** [review]

**Pull request merged**